crates/cosmos/cosmos-chain-components/src/impls/misbehaviour/check_misbehaviour.rs

```diff
@@ -0,0 +1,285 @@
+                        header
+                            .trusted_height
+                            .ok_or_else(|| {
+                                Chain::raise_error("`trusted_height` missing from support `Header`")
+                            })?
+                            .revision_height
+                            + 1,
+                    )
+                    .map_err(Chain::raise_error)?;
+
+                    // Therefore we can now trust the next validator set, see NOTE above.
+                    current_trusted_validators = light_client
+                        .fetch_light_block(&next_height)
+                        .await
+                        .map_err(Chain::raise_error)?
+                        .validators;
+
+                    supporting_headers.push(header);
+                }
+
+                // a) Set the trusted height of the target header to the height of the previous
+                // supporting header if any, or to the initial trusting height otherwise.
+                //
+                // b) Set the trusted validators of the target header to the validators of the successor to
+                // the last supporting header if any, or to the initial trusted validators otherwise.
+                let (latest_trusted_height, latest_trusted_validator_set) =
+                    match supporting_headers.last() {
+                        Some(prev_header) => {
+                            let prev_height = TendermintHeight::try_from(
+                                prev_header
+                                    .trusted_height
+                                    .ok_or_else(|| {
+                                        Chain::raise_error(
+                                            "`trusted_height` missing from previous `Header`",
+                                        )
+                                    })?
+                                    .revision_height
+                                    + 1,
+                            )
+                            .map_err(Chain::raise_error)?;
+                            let prev_succ = light_client
+                                .fetch_light_block(&prev_height)
+                                .await
+                                .map_err(Chain::raise_error)?;
+                            (
+                                prev_header.trusted_height.ok_or_else(|| {
+                                    Chain::raise_error(
+                                        "`trusted_height` missing from previous `Header`",
+                                    )
+                                })?,
+                                prev_succ.validators,
+                            )
+                        }
+                        None => (trusted_height.into(), trusted_validator_set),
+                    };
+
+                #[allow(deprecated)]
+                Ok(Some(Misbehaviour {
```

Comment:
hmm. interesting that we are using Protobuf types directly.

Reply:
shouldn't the domain have the necessary conversion to proto encoding ?

if it doesn't, ok to use protobuf type here.

Reply:
No domain type was created, I will put a TODO